### PR TITLE
More orm relationships

### PR DESCRIPTION
**Pull request by chrisbay**

### CODE REVIEW

**Comment by chrisbay** [comment]
[EPIC - Java Lesson 18 🎆](https://trello.com/c/GRByPWcF/372-epic-java-lesson-18-%F0%9F%8E%86)

File: src/chapters/orm-relationships/one-to-one.rst [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+   @NotNull
+   private EventDetails eventDetails;
+
+.. index:: @Valid
+
+You have used ``@NotNull`` in this way before, but this is the first time we have used ``@Valid`` on a class member. 
+
+As we have seen, using ``@Valid`` on a method parameter in a controller will result in the fields of that method being validated. For instance, with an ``Event`` object, our ``@NotNull`` annotation will ensure that the ``eventDetails`` field is not null. By default, however, validation will not descend into the ``eventDetails`` class to check *its* validation annotations. 
+
+Using ``@Valid`` on the ``eventDetails`` field ensures that such validation occurs. It makes sure that an ``Event`` object will not be considered valid unless it has an ``EventDetails`` object that is *also* valid (i.e. it has valid ``description`` and ``contactEmail`` fields).
```
Comment (CarlyLanglois) [inline src/chapters/orm-relationships/one-to-one.rst]: I got a little twisted around reading these paragraphs, forgetting which class we were looking at, so I tried rewording it:

"This is the first time we have used ``@Valid`` on a class member. Let's explore what it is doing here.

First, let's review what ``@NotNull`` accomplishes. When an ``Event`` object is created, the ``@NotNull`` annotation will ensure that the ``eventDetails`` field 
is not null. But what if we want to ensure that ``eventDetails`` is itself a valid object as the validation process for ``Event``? By default, model validation will not descend 
into the ``eventDetails`` class to check *its* validation annotations. However, annotating the field with ``@Valid`` overrides the default behavior. It makes sure that an 
``Event`` object will not be considered valid unless it has an ``EventDetails`` object that is *also* valid (i.e. it has valid ``description`` and ``contactEmail`` fields)."

Not crucial to change though

File: src/chapters/orm-relationships/one-to-one.rst [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+
+.. index:: ! transient
+
+This exception refers to the transient value ``Event.eventDetails``. A **transient** value is a an object that *can* be saved to the database (i.e. is of an entity type) but has NOT been saved yet. In our case, trying to save ``newEvent`` causes problems because its ``eventDetails`` field can not be null in the database, but the value of this field---a new ``EventDetails`` object created on form submission---has not been saved yet.
+
+.. index:: ! cascade
+
+The fix for this problem is simple, and allows us to introduce the concept of cascading. A database operation **cascades** from ``Event`` to ``EventDetails`` if, when the operation is applied to an ``Event`` instance, it is also applied to the associated ``EventDetails`` instance. If our call to ``eventRepository.save`` could be made to *cascade* then our problem would be solved! 
+
+To force our save operation to cascade, go into the ``Event`` class and add a ``cascade`` parameter to the ``@OneToOne`` annotation:
```
Comment (CarlyLanglois) [inline src/chapters/orm-relationships/one-to-one.rst]: personal preference, I'd drop the 'force' from this construction.

"To cascade our ``eventRepository.save`` operation, ..."

File: src/chapters/orm-relationships/one-to-one.rst [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+We *could* set ``cascade = CascadeType.PERSIST`` and solve our current problem as well. However, that would mean that delete operations would not cascade. It makes sense for the ``EventDetails`` object to be deleted when its associated ``Event`` object is deleted, so we use ``CascadeType.ALL``.
+
+As you continue working with ORM, you are likely to need to use other ``CascadeType`` values. We won't go into more depth on this topic here, but encourage you to `read the documentation <https://docs.oracle.com/javaee/6/api/javax/persistence/CascadeType.html>`_ on your own.
+
+At this point, your app should work. We have established our first one-to-one relationship, while learning about a new design pattern and cascading. Nice work! 
+
+The Inverse Relationship
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+Once we have set up the relationship from ``Event`` to ``EventDetails`` it is easy to configure the inverse relationship.
```
Comment (CarlyLanglois) [inline src/chapters/orm-relationships/one-to-one.rst]: add a comma after ``EventDetails``

File: src/chapters/orm-relationships/one-to-one.rst [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+
+To do so, add a field of type ``Event`` to ``EventDetails``. Then add ``@OneToOne`` to the new field with a ``mappedBy`` parameter.
+
+.. sourcecode:: java
+
+   @OneToOne(mappedBy = "eventDetails")
+   private Event event;
+
+Setting ``mappedBy = "eventDetails"`` will ensure that the field is populated correctly. For a specific ``EventDetails`` object ``details``, ``event`` will be populated with the ``Event`` object that contains ``details``. Then both sides of the one-to-one relationship will have a reference to the other.
```
Comment (CarlyLanglois) [inline src/chapters/orm-relationships/one-to-one.rst:216]: In your video segment, you show creating the inverse relationship and then remove it, mentioning that this application does not have this inverse relationship btn event and event details. its not a big deal, but you might want to mention something in this section, like - 
"If you wanted to add the inverse relationship, this is what is done and you can use the mappedBy param on oneToOne to set up efficient table mapping. The inverse relationship is not required, however, for this application."

File: src/chapters/orm-relationships/one-to-one.rst [hunk 5]
```diff
@@ -0,0 +1,232 @@
+Creating a One-to-One Relationship
+==================================
+
+We :ref:`defined a one-to-one relationship <one-to-one-def>` between two objects, A and B, as occurring when an object of type A can be related to only one instance of an object of type B, and vice versa.
+
+Such a relationship can be configured using the JPA annotation ``@OneToOne``.
+
+Creating a One-to-One Relationship - Video
```
Comment (CarlyLanglois) [inline src/chapters/orm-relationships/one-to-one.rst:8]: I don't think you want to change it now, but just fyi the video starts in what sounds like the middle of a sentence - or the middle of a thought. We're missing the "hi this is chris.." intro but it also is not an illogical start. everything makes sense as its cut now.

File: src/chapters/orm-relationships/one-to-one.rst [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+Setting ``mappedBy = "eventDetails"`` will ensure that the field is populated correctly. For a specific ``EventDetails`` object ``details``, ``event`` will be populated with the ``Event`` object that contains ``details``. Then both sides of the one-to-one relationship will have a reference to the other.
+
+Check Your Understanding
+------------------------
+
+.. admonition:: Question
+
+   True/False: When a new object is saved to a repository, all of its non-primitive fields are saved as well.
+
+.. ans: False. Any fields that are also entities must be either explicitly, or else the appropriate cascade setting must be used.
```
Comment (CarlyLanglois) [inline src/chapters/orm-relationships/one-to-one.rst]: explicitly saved as well?

File: src/chapters/orm-relationships/many-to-many.rst [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+====================================
+
+Just as when introducing one-to-one relationships, our first step in exploring many-to-many relationships is to introduce a new class to relate to ``Event``.
+
+Persistent Tags - Video
+-----------------------
+
+.. raw:: html
+
+   <div style="text-align:center;"><iframe width="800" height="450" src="https://www.youtube.com/embed/GKOCCjn86yk" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe></div>
```
Comment (CarlyLanglois) [inline src/chapters/orm-relationships/many-to-many.rst]: oups! this is the wrong video - I think you want 6.2

https://www.youtube.com/watch?v=bLK-VtZgx0Q

Comment (CarlyLanglois) [reply src/chapters/orm-relationships/many-to-many.rst]: and that video looks good - and lol I wrote the exercises to have the students add this exact same class 😄

File: src/chapters/orm-relationships/many-to-many.rst [hunk 8]
```diff
@@ -0,0 +1,332 @@
...
+   <div style="text-align:center;"><iframe width="800" height="450" src="https://www.youtube.com/embed/1qMaEv_CJ6k" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe></div>
+
+The final code from this video is in the `many-to-many-features branch <https://github.com/LaunchCodeEducation/coding-events/tree/many-to-many-features>`__ of ``coding-events``.
+
+Many-to-Many Forms and Data Transfer Objects - Text
+---------------------------------------------------
+
+.. index:: design pattern
+
+In order to make use of our new many-to-many relationship, we need to have a way for users to add a tag to an event. We will implement this feature by creating a form that, for a given event, let's the user add a single tag. The process of binding an ``Event`` and ``Tag`` object together will be made easier through the use of a new design pattern.
```
Comment (CarlyLanglois) [inline src/chapters/orm-relationships/many-to-many.rst]: lets* no apostrophe

Comment (CarlyLanglois) [reply src/chapters/orm-relationships/many-to-many.rst]: an ``Event`` and a ``Tag``

File: src/chapters/orm-relationships/many-to-many.rst [hunk 9]
```diff
@@ -0,0 +1,332 @@
...
+      EventTagDTO eventTag = new EventTagDTO();
+      eventTag.setEvent(event);
+      model.addAttribute("eventTag", eventTag);
+      return "events/add-tag.html";
+   }
+
+- **Line 112**: Specifies that the handler will be available at the route ``/events/add-tag``, and will respond to ``GET`` requests.
+- **Line 113**: Defines the ``displayAddTagForm`` handler, which has a required query parameter, ``eventId``.
+- **Line 114**: Queries the repository for the ``Event`` object with ID equal to the value of ``eventId``.
+- **Line 115**: Extracts the ``Event`` object from the result of the query. We would ideally include a conditional to check that such an object exists before proceeding.
```
Comment (CarlyLanglois) [inline src/chapters/orm-relationships/many-to-many.rst]: We can include some language explaining why this is not included here (for space, for focus on the DTO usage?)

File: src/chapters/orm-relationships/many-to-many.rst [hunk 10]
```diff
@@ -0,0 +1,332 @@
...
+Lines 130-131 retrieve the values of these fields from the DTO. Then, as long as the given event doesn't already have the given tag, we add the tag to it's collection in lines 132-134. Finally, we save the ``event`` to ``eventRepository``, which results in the relationship being stored in the database.
+
+Exactly *how* this relationship is stored utilizes a new type of SQL table.
+
+Join Tables
+^^^^^^^^^^^
+
+Think about how relationships are established at the database level. One-to-one and one-to-many relationships are facilitated by the use of a foreign key column on one side of the relationship. Our ``event`` table has two foreign key columns: ``event_category_id`` and ``event_details_id``. 
+
+For a given row in ``event``, the column ``event_category_id`` contains the primary key of the row in ``event_category`` that the ``event`` row is related to, an similarly for ``event_details_id``.
```
Comment (CarlyLanglois) [inline src/chapters/orm-relationships/many-to-many.rst]: and similarly*

File: src/chapters/orm-relationships/many-to-many.rst [hunk 11]
```diff
@@ -0,0 +1,332 @@
...
+Exactly *how* this relationship is stored utilizes a new type of SQL table.
+
+Join Tables
+^^^^^^^^^^^
+
+Think about how relationships are established at the database level. One-to-one and one-to-many relationships are facilitated by the use of a foreign key column on one side of the relationship. Our ``event`` table has two foreign key columns: ``event_category_id`` and ``event_details_id``. 
+
+For a given row in ``event``, the column ``event_category_id`` contains the primary key of the row in ``event_category`` that the ``event`` row is related to, an similarly for ``event_details_id``. 
+
+The only difference is the *number* of different ``event`` rows that may have the same value of ``event_category_id`` and ``event_details_id``. The ``event``/``event_category`` relationship is many-to-one, so *many* event rows may have the same ``event_category_id`` value. The ``event``/``event_details`` relationship is one-to-one, so *only one* event row may have a given value in ``event_details_id``.
```
Comment (CarlyLanglois) [inline src/chapters/orm-relationships/many-to-many.rst]: I think replacing the use of "given" in the last sentence with "certain" might make this point clearer that it has to do with a specific value, not just having a provided value

File: src/chapters/orm-relationships/many-to-many.rst [hunk 12]
```diff
@@ -0,0 +1,332 @@
...
+
+For a given row in ``event``, the column ``event_category_id`` contains the primary key of the row in ``event_category`` that the ``event`` row is related to, an similarly for ``event_details_id``. 
+
+The only difference is the *number* of different ``event`` rows that may have the same value of ``event_category_id`` and ``event_details_id``. The ``event``/``event_category`` relationship is many-to-one, so *many* event rows may have the same ``event_category_id`` value. The ``event``/``event_details`` relationship is one-to-one, so *only one* event row may have a given value in ``event_details_id``. 
+
+.. index:: ! join table
+
+Using foreign and primary keys to create many-to-many relationships is a bit trickier. In order to relate rows in ``event`` to rows in ``tag`` we need need a third table, known as a **join table**. A join table consists of two columns, each of which is a foreign key column to another table. Each row in a join table represents a relationship between one row of each of the two tables. This technique enables many-to-many relationships.
```
Comment (CarlyLanglois) [inline src/chapters/orm-relationships/many-to-many.rst]: for the second to last sentence, what do you think of the following?  
"Each row in a join table represents a relationship between one row **in** each of the two tables."
Just using in rather than of

File: src/chapters/orm-relationships/many-to-many.rst [hunk 13]
```diff
@@ -0,0 +1,332 @@
...
+
+.. list-table:: A join table with a single relationship
+   :header-rows: 1
+
+   * - events_id
+     - tags_id
+   * - 13
+     - 4
+
+We can do this again and again to generate more relationships. Let's revisit the :ref:`many-to-many diagram <many-to-many-figure>` from earlier in the chapter.
```
Comment (CarlyLanglois) [inline src/chapters/orm-relationships/many-to-many.rst]: if it is the same figure, it seems counter-productive to link back to it. I'd remove the ref or change the ref to the section heading that introduces the many to many relationship in the chapter

File: src/chapters/orm-relationships/many-to-many.rst [hunk 14]
```diff
@@ -0,0 +1,332 @@
...
+.. ans: False. Any class may be used with model binding
+
+.. admonition:: Question
+
+   The use of join tables enables (select all that apply):
+
+   #. A database where you never need to run a ``JOIN`` query.
+   #. Many-to-many relationships between tables.
+   #. Many-to-many relationships between classes without using the ``@ManyToMany`` annotation.
+   #. Rainbows and butterflies to be stored in your database.
```
Comment (CarlyLanglois) [inline src/chapters/orm-relationships/many-to-many.rst:334]: ans: b - only b I think?

File: src/chapters/orm-relationships/many-to-many.rst [hunk 15]
```diff
@@ -0,0 +1,332 @@
...
+               event.addTag(tag);
+               eventRepository.save(event);
+         }
+         return "redirect:detail?eventId=" + event.getId();
+      }
+
+      return "redirect:add-tag";
+   }
+
+Using model binding, our method takes a valid parameter of type ``EventTagDTO``. Since we referenced the ``event`` and ``tag`` fields of our DTO when rendering the form (in the template, using ``th:field``), our submitted form should contain all of the data necessary to create an ``EventTagDTO`` instance. This instance will be valid if both ``event`` and ``tag`` are non-null.
```
Comment (CarlyLanglois) [inline src/chapters/orm-relationships/many-to-many.rst:215]: Like we talked about on slack, i'd append this paragraph reinforcing why DTOs are handy - your words:
"The DTO basically allows us to have everything automatically created and validated using model binding. Otherwise, we'd have to manually grab an event ID and a tag ID, query the repositories, validate the results, and so on."

I'd rephrase to something like:

"The reasons for creating a DTO model should hopefully be a bit clearer by now. Using a DTO allows us to create and validate these objects through model binding. The same event and tag relationship information could be processed without a DTO, but this would require passing query parameters for the ids of both event and tag objects, querying the Event Repo and TagRepo for these items, validating those objects, etc. Simply put, the DTO makes this procedure a little cleaner and easier."

**Review by CarlyLanglois — APPROVED** [review]

File: src/chapters/orm-relationships/one-to-one.rst (hunk 5 above)
Comment (chrisbaylewis) [reply src/chapters/orm-relationships/one-to-one.rst:8]: I noticed that, but didn't have a way to fix. I'll re-record all of these videos later anyway, since I used the wrong microphone.
